### ISSUE TITLE
Crash on ARM

### ISSUE THREAD

**Issue by ViralBShah** [FFTW/fftw3#30]
This FFTW test crashes on ARM. Running on a chromebook, with an ubuntu 12.04 chroot. GCC version is 4.8.2.

````
make[3]: Entering directory `/home/viral/julia/deps/fftw-3.3.4-single/tests'
perl -w ./check.pl  -r -c=30 -v `pwd`/bench
Executing "/home/viral/julia/deps/fftw-3.3.4-single/tests/bench --verbose=1   --verify 'ok88o01*152' --verify 'ik88o01*152' --verify 'obr5x14x2x11*5' --verify 'ibr5x14x2x11*5' --verify 'ofr5x14x2x11*5' --verify 'ifr5x14x2x11*5' --verify '//obc5x14x2x11*5' --verify '//ibc5x14x2x11*5' --verify '//ofc5x14x2x11*5' --verify '//ifc5x14x2x11*5' --verify 'obc5x14x2x11*5' --verify 'ibc5x14x2x11*5' --verify 'ofc5x14x2x11*5' --verify 'ifc5x14x2x11*5' --verify 'ok8e01*29' --verify 'ik8e01*29' --verify 'obr4x2v15' --verify 'ibr4x2v15' --verify 'ofr4x2v15' --verify 'ifr4x2v15' --verify '//obc4x2v15' --verify '//ibc4x2v15' --verify '//ofc4x2v15' --verify '//ifc4x2v15' --verify 'obc4x2v15' --verify 'ibc4x2v15' --verify 'ofc4x2v15' --verify 'ifc4x2v15' --verify 'ok8e10x7bx11e00x10e10' --verify 'ik8e10x7bx11e00x10e10' --verify '//obr40x11' --verify '//ofr40x11' --verify 'obr40x11' --verify 'ibr40x11' --verify 'ofr40x11' --verify 'ifr40x11' --verify '//obc40x11' --verify '//ibc40x11' --verify '//ofc40x11' --verify '//ifc40x11' --verify 'obc40x11' --verify 'ibc40x11' --verify 'ofc40x11' --verify 'ifc40x11'"
Segmentation fault (core dumped)
````

**Comment by matteo-frigo** [comment]
It's probably the NEON autodetection code.  Try editing
simd-support/neon.c, replacing really_have_neon() with "return 1;"

Any suggestion on how to do this autodetection correctly is most welcome.

**Comment by ViralBShah** [comment]
Yes, that fixes it. I don't have any idea how to do the autodetection correctly though. Just thinking out aloud, autoconf seems to have some NEON detection code. Just wondering if that is useful in any way.

If I disable NEON during the FFTW configure, would I not face this problem?

**Comment by stevengj** [comment]
We might look at how the [https://github.com/steinwurf/cpuid](cpuid library) does it.  In particular, it just checks [#ifdef __ARM_NEON__](https://github.com/steinwurf/cpuid/blob/693e94e02ebd9e7943eaaab4560c6d53e1dd2c2c/src/cpuid/init_ios_clang_arm.hpp#L22) when compiled with Clang, and [reads "/proc/self/auxv"](https://github.com/steinwurf/cpuid/blob/693e94e02ebd9e7943eaaab4560c6d53e1dd2c2c/src/cpuid/init_linux_gcc_arm.hpp) on Linux.

**Comment by stevengj** [comment]
Cairo may also have some neon autodetection.

**Comment by matteo-frigo** [comment]
"Viral B. Shah" <notifications@github.com> writes:

> Yes, that fixes it. I don't have any idea how to do the autodetection
> correctly though. Just thinking out aloud, autoconf seems to have some
> NEON detection code. Just wondering if that is useful in any way.
>

Viral,

does replacing

     asm volatile (".long 0xf2000150");

with
     asm volatile ("vand q0, q0, q0");

work?

I don't like the idea of reading /proc/self/auxv because it is
linux-specific (then the BSD and iOS folks will correctly complain).

I don't like the idea of a compile-time constant, because this will make
life harder for e.g. Debian which is supposed to run on systems with and
without NEON, and then Debian would have to ship two separate binaries.

**Comment by ViralBShah** [comment]
That doesn't work:

````
/tmp/cc3B6tIO.s: Assembler messages:
/tmp/cc3B6tIO.s:36: Error: selected processor does not support Thumb mode `vand q0,q0,q0'
````

**Comment by stevengj** [comment]
(@matteo-frigo, Cairo uses a compile-time constant on iOS, something Android-specific on Android, and `/proc/self/auxv` on Linux.  But then it doesn't appear to work with BSD, of course.)

**Comment by rdolbeau** [comment]
I have the same segmentation fault with Ubuntu's default 4.8.2. However, it works with the Ubuntu's default 4.7.. It also works with self-compiled FSF GCC 4.8.4, 4.9.1, 4.9.2 and 5.1.0-RC-20150412.
Seems like a compiler bug in 4.8.2.

**Comment by vtjnash** [comment]
that instruction is only allowed when the processor is in privileged mode, so you can't use it in user mode (unlike cpuid on x86).

the preferred method appears to be to parse the `/proc/cpuinfo` for the `Features:` string for linux. For iOS, the preferred method is to use the compiler flag (and assume the compiler will build a fat binary). It looks like the cairo code actually just executes a neon instruction and catches the possible resulting SIGILL?

**Comment by rdolbeau** [comment]
@vtjnash: which instruction? "vand" is a Neon instruction, and available in userland.
And yes, the code detects Neon at runtime by detecting SIGILL when using a Neon instruction.
Portability is an issue, so per-OS solutions are not favored...
